File: omniaquant-apex/src/gtm/codec.rs
//! GTM Codec — Orchestrates the full GTM encode/decode pipeline:
//! RHT → Polar Transform → Lloyd-Max Quantization → QJL Bias Correction

use ndarray::Array1;
use serde::{Deserialize, Serialize};

use super::polar::{polar_inverse, polar_transform};
use super::qjl::QJL;
use super::quantize::LloydMaxQuantizer;
use super::rht::RHT;

const CHUNK_SIZE: usize = 32;

/// Serializable container for a quantized latent chunk.
#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct GTMPacket {
    pub chunk_idx: usize,
    pub chunk_size: usize,
    pub r_idx: u16,
    pub theta_indices: Vec<u16>,
    pub qjl_bits: Vec<bool>,
    pub n_bits_r: usize,
    pub n_bits_theta: usize,
    pub padded_dim: usize,
}

/// GTM Encoder: applies RHT + polar quantization + QJL to a latent vector.
pub struct GTMEncoder {
    n_bits: usize,
    n_bits_r: usize,
    n_bits_theta: usize,
    qjl: QJL,
    qz_r: LloydMaxQuantizer,
    qz_theta: LloydMaxQuantizer,
    rhts: Vec<(usize, u64, RHT)>,
}

impl GTMEncoder {
    pub fn new(n_bits: usize, qjl_proj_dim: usize, seed: u64) -> Self {
        let n_bits_r = std::cmp::min(n_bits + 2, 8);
        let n_bits_theta = n_bits;

        let mut qz_r = LloydMaxQuantizer::new(n_bits_r);
        qz_r.default_fit(3.0);

        let mut qz_theta = LloydMaxQuantizer::new(n_bits_theta);
        qz_theta.default_fit(std::f64::consts::FRAC_PI_2);

        Self {
            n_bits,
            n_bits_r,
            n_bits_theta,
            qjl: QJL::new(qjl_proj_dim, seed.wrapping_add(1)),
            qz_r,
            qz_theta,
            rhts: Vec::new(),
        }
    }

    fn get_rht(&mut self, dim: usize, seed_offset: usize) -> &RHT {
        let seed = 42u64.wrapping_add(seed_offset as u64);
        if !self.rhts.iter().any(|(d, s, _)| *d == dim && *s == seed) {
            self.rhts.push((dim, seed, RHT::new(dim, seed)));
        }
        let idx = self
            .rhts
            .iter()
            .position(|(d, s, _)| *d == dim && *s == seed)
            .unwrap();
        &self.rhts[idx].2
    }

    /// Encode a 1-D latent vector into a list of GTMPackets.
    pub fn encode(&mut self, v: &Array1<f64>) -> Vec<GTMPacket> {
        let d = v.len();
        let n_chunks = (d + CHUNK_SIZE - 1) / CHUNK_SIZE;
        let mut packets = Vec::with_capacity(n_chunks);

        for i in 0..n_chunks {
            let start = i * CHUNK_SIZE;
            let end = std::cmp::min(start + CHUNK_SIZE, d);
            let chunk = v.slice(ndarray::s![start..end]).to_owned();
            let actual_size = chunk.len();

            // Build RHT separately to avoid borrow conflicts
            let seed = 42u64.wrapping_add(i as u64);
            let rht = RHT::new(actual_size, seed);
            let v_rht = rht.forward(&chunk);

            // Polar decompose
            let (r, thetas) = polar_transform(v_rht.view());

            // Quantize radius
            let r_idx = self.qz_r.quantize(r);

            // Quantize angles
            let theta_indices: Vec<u16> =
                thetas.iter().map(|&t| self.qz_theta.quantize(t)).collect();

            // Reconstruct for QJL residual
            let r_rec = self.qz_r.dequantize(r_idx);
            let theta_rec: Array1<f64> = theta_indices
                .iter()
                .map(|&idx| self.qz_theta.dequantize(idx))
                .collect();
            let v_rec = polar_inverse(r_rec, &theta_rec);
            let v_inv = rht.inverse(&v_rec);
            let residual = &chunk - &v_inv;

            // QJL encode residual
            let qjl_signs = self.qjl.encode(&residual);

            packets.push(GTMPacket {
                chunk_idx: i,
                chunk_size: actual_size,
                r_idx,
                theta_indices,
                qjl_bits: qjl_signs,
                n_bits_r: self.n_bits_r,
                n_bits_theta: self.n_bits_theta,
                padded_dim: rht.padded_dim(),
            });
        }

        packets
    }

    /// Encode then immediately decode — used for LCC preview.
    pub fn encode_decode(&mut self, v: &Array1<f64>) -> Array1<f64> {
        let packets = self.encode(v);
        GTMDecoder::new(self.qjl.proj_dim()).decode(&packets, v.len())
    }
}

/// GTM Decoder: reconstructs a latent vector from GTMPackets.
pub struct GTMDecoder {
    qjl: QJL,
    rhts: Vec<(usize, u64, RHT)>,
    qz_cache: std::collections::HashMap<usize, LloydMaxQuantizer>,
}

impl GTMDecoder {
    pub fn new(qjl_proj_dim: usize) -> Self {
        Self {
            qjl: QJL::new(qjl_proj_dim, 0),
            rhts: Vec::new(),
            qz_cache: std::collections::HashMap::new(),
        }
    }

    fn get_rht(&mut self, dim: usize, seed_offset: usize) -> &RHT {
        let seed = 42u64.wrapping_add(seed_offset as u64);
        if !self.rhts.iter().any(|(d, s, _)| *d == dim && *s == seed) {
            self.rhts.push((dim, seed, RHT::new(dim, seed)));
        }
        let idx = self
            .rhts
            .iter()
            .position(|(d, s, _)| *d == dim && *s == seed)
            .unwrap();
        &self.rhts[idx].2
    }

    fn get_qz_theta(&mut self, n_bits: usize) -> &LloydMaxQuantizer {
        self.qz_cache.entry(n_bits).or_insert_with(|| {
            let mut qz = LloydMaxQuantizer::new(n_bits);
            qz.default_fit(std::f64::consts::FRAC_PI_2);
            qz
        })
    }

    /// Reconstruct latent vector from packets.
    pub fn decode(&mut self, packets: &[GTMPacket], orig_dim: usize) -> Array1<f64> {
        let mut out = Array1::zeros(orig_dim);
        let mut sorted = packets.to_vec();
        sorted.sort_by_key(|p| p.chunk_idx);

        for pkt in &sorted {
            let i = pkt.chunk_idx;

            // Get RHT parameters first (clone to avoid borrow conflict)
            let seed = 42u64.wrapping_add(i as u64);
            let rht = RHT::new(pkt.chunk_size, seed);

            let mut qz_r = LloydMaxQuantizer::new(pkt.n_bits_r);
            qz_r.default_fit(3.0);

            // Build theta quantizer
            let mut qz_theta = LloydMaxQuantizer::new(pkt.n_bits_theta);
            qz_theta.default_fit(std::f64::consts::FRAC_PI_2);

            let r_rec = qz_r.dequantize(pkt.r_idx);
            let theta_rec: Array1<f64> = pkt
                .theta_indices
                .iter()
                .map(|&idx| qz_theta.dequantize(idx))
                .collect();

            let v_polar = polar_inverse(r_rec, &theta_rec);
            let v_chunk = rht.inverse(&v_polar);

            // QJL correction
            let correction = self.qjl.decode(&pkt.qjl_bits, pkt.chunk_size);
            let v_corrected = v_chunk + correction;

            let start = i * CHUNK_SIZE;
            let end = std::cmp::min(start + pkt.chunk_size, orig_dim);
            for (j, k) in (start..end).enumerate() {
                out[k] = v_corrected[j];
            }
        }

        out
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_gtm_round_trip() {
        let mut enc = GTMEncoder::new(4, 32, 42);
        let v: Array1<f64> = Array1::from_vec((0..64).map(|i| (i as f64 - 32.0) * 0.1).collect());

        let packets = enc.encode(&v);
        let mut dec = GTMDecoder::new(32);
        let v_rec = dec.decode(&packets, 64);

        let err = (&v - &v_rec).mapv(|x| x * x).sum().sqrt();
        assert!(err < v.mapv(|x| x * x).sum().sqrt() * 0.8);
    }

    #[test]
    fn test_packet_serialization() {
        let mut enc = GTMEncoder::new(4, 32, 42);
        let v: Array1<f64> = Array1::from_vec((0..64).map(|i| i as f64 * 0.05).collect());
        let packets = enc.encode(&v);

        for pkt in &packets {
            let json = serde_json::to_string(pkt).unwrap();
            let pkt2: GTMPacket = serde_json::from_str(&json).unwrap();
            assert_eq!(pkt2.chunk_idx, pkt.chunk_idx);
            assert_eq!(pkt2.r_idx, pkt.r_idx);
            assert_eq!(pkt2.theta_indices, pkt.theta_indices);
        }
    }
}

File: gtm/codec.py
"""
GTM Codec — Orchestrates the full GTM encode/decode pipeline:
RHT → Polar Transform → Lloyd-Max Quantization → QJL Bias Correction

Supports:
  - Full vector encoding (keyframe latents z_t)
  - Chunked processing for large D
  - Adaptive bit allocation (more bits for radius, fewer for angles)
  - Serializable GTMPackets for transmission over a network channel
"""
import torch
import math
from dataclasses import dataclass, field
from typing import List, Optional, Tuple
import io, struct

from .rht import RHT
from .polar import polar_transform, polar_inverse
from .quantize import LloydMaxQuantizer
from .qjl import QJL


@dataclass
class GTMPacket:
    """Serialisable container for a quantized latent chunk."""
    chunk_idx: int          # chunk index within full vector
    chunk_size: int         # original chunk dimensionality
    r_idx: int              # quantized radius index
    theta_indices: List[int]  # quantized angle indices
    qjl_bits: bytes         # packed QJL sign bits
    n_bits_r: int
    n_bits_theta: int
    padded_dim: int

    def to_bytes(self) -> bytes:
        buf = io.BytesIO()
        # !HHHHHHi = 5*2 + 4 + 4 = 18 bytes total
        buf.write(struct.pack("!HHHHHHi", self.chunk_idx, self.chunk_size,
                              self.n_bits_r, self.n_bits_theta,
                              len(self.theta_indices), self.padded_dim, self.r_idx))
        for t in self.theta_indices:
            buf.write(struct.pack("!H", t))
        buf.write(struct.pack("!H", len(self.qjl_bits)))
        buf.write(self.qjl_bits)
        return buf.getvalue()

    @classmethod
    def from_bytes(cls, data: bytes) -> "GTMPacket":
        buf = io.BytesIO(data)
        # !HHHHHHi = 16 bytes (5*H + I + i)
        header = buf.read(16)
        if len(header) < 16:
            raise ValueError(f"GTMPacket header incomplete: {len(header)}/16 bytes")
        chunk_idx, chunk_size, n_bits_r, n_bits_theta, n_theta, padded_dim, r_idx = \
            struct.unpack("!HHHHHHi", header)
        theta_indices = []
        for _ in range(n_theta):
            t_bytes = buf.read(2)
            if len(t_bytes) < 2:
                raise ValueError("Incomplete theta index")
            theta_indices.append(struct.unpack("!H", t_bytes)[0])
        qjl_len_bytes = buf.read(2)
        if len(qjl_len_bytes) < 2:
            raise ValueError("Missing QJL length")
        qjl_len = struct.unpack("!H", qjl_len_bytes)[0]
        qjl_bits = buf.read(qjl_len)
        if len(qjl_bits) < qjl_len:
            raise ValueError(f"Incomplete QJL bits: {len(qjl_bits)}/{qjl_len}")
        return cls(chunk_idx=chunk_idx, chunk_size=chunk_size, r_idx=r_idx,
                   theta_indices=theta_indices, qjl_bits=qjl_bits,
                   n_bits_r=n_bits_r, n_bits_theta=n_bits_theta, padded_dim=padded_dim)


class GTMEncoder:
    """
    Applies RHT + polar quantization + QJL to a latent vector.
    Produces a list of GTMPacket objects (one per chunk).
    """

    CHUNK_SIZE = 32  # process in 32-dim chunks

    def __init__(self, n_bits: int = 4, qjl_proj_dim: int = 64,
                 n_bits_r: Optional[int] = None, seed: int = 42):
        self.n_bits = n_bits
        self.n_bits_r = n_bits_r or min(n_bits + 2, 8)  # radius gets extra bits
        self.n_bits_theta = n_bits
        self.qjl = QJL(proj_dim=qjl_proj_dim, seed=seed + 1)
        self._qz_r = LloydMaxQuantizer(self.n_bits_r).default_fit(scale=3.0)
        self._qz_theta = LloydMaxQuantizer(self.n_bits_theta).default_fit(scale=math.pi / 2)
        self._rhts: dict = {}  # chunk_size → RHT

    def _get_rht(self, dim: int, seed_offset: int) -> RHT:
        key = (dim, seed_offset)
        if key not in self._rhts:
            self._rhts[key] = RHT(dim, seed=42 + seed_offset)
        return self._rhts[key]

    def encode(self, v: torch.Tensor) -> List[GTMPacket]:
        """
        Encode 1-D latent vector v (shape: D,) into a list of GTMPackets.
        """
        packets = []
        D = v.shape[0]
        cs = self.CHUNK_SIZE
        n_chunks = math.ceil(D / cs)

        for i in range(n_chunks):
            chunk = v[i * cs: (i + 1) * cs]
            actual_size = chunk.shape[0]

            rht = self._get_rht(actual_size, i)
            v_rht = rht.forward(chunk)                       # (padded_dim,)

            # Polar decompose
            r, thetas = polar_transform(v_rht.unsqueeze(0))  # (1,), (1, pd-1)
            r = r.squeeze(0)
            thetas = thetas.squeeze(0)

            # Quantize radius
            r_idx = int(self._qz_r.quantize(r.unsqueeze(0)).item())

            # Quantize angles
            theta_indices = self._qz_theta.quantize(thetas).tolist()

            # Reconstruct to compute residual for QJL
            r_rec = self._qz_r.dequantize(torch.tensor([r_idx]))
            theta_rec = self._qz_theta.dequantize(torch.tensor(theta_indices))
            v_rec = polar_inverse(r_rec, theta_rec.unsqueeze(0)).squeeze(0)
            v_original_back = rht.inverse(v_rht.unsqueeze(0) if False else v_rec.unsqueeze(0)).squeeze(0)
            residual = chunk - rht.inverse(v_rec.unsqueeze(0)).squeeze(0)

            # QJL encode residual
            qjl_signs = self.qjl.encode(residual)           # (proj_dim,) bool
            qjl_bytes = _pack_bits(qjl_signs)

            packets.append(GTMPacket(
                chunk_idx=i, chunk_size=actual_size,
                r_idx=r_idx, theta_indices=[int(t) for t in theta_indices],
                qjl_bits=qjl_bytes,
                n_bits_r=self.n_bits_r, n_bits_theta=self.n_bits_theta,
                padded_dim=rht.padded_dim,
            ))

        return packets

    def encode_decode(self, v: torch.Tensor) -> torch.Tensor:
        """Encode then immediately decode — used for LCC preview."""
        packets = self.encode(v)
        return GTMDecoder(qjl_proj_dim=self.qjl.proj_dim).decode(packets, v.shape[0])


class GTMDecoder:
    """Reconstructs a latent vector from a list of GTMPackets."""

    def __init__(self, qjl_proj_dim: int = 64, seed: int = 42):
        self.qjl = QJL(proj_dim=qjl_proj_dim, seed=seed + 1)
        self._rhts: dict = {}
        self._qz_cache: dict = {}

    def _get_rht(self, dim: int, seed_offset: int) -> RHT:
        key = (dim, seed_offset)
        if key not in self._rhts:
            self._rhts[key] = RHT(dim, seed=42 + seed_offset)
        return self._rhts[key]

    def _get_qz(self, n_bits: int) -> LloydMaxQuantizer:
        if n_bits not in self._qz_cache:
            self._qz_cache[n_bits] = LloydMaxQuantizer(n_bits).default_fit(scale=math.pi / 2)
        return self._qz_cache[n_bits]

    def decode(self, packets: List[GTMPacket], orig_dim: int) -> torch.Tensor:
        """Reconstruct latent vector from packets."""
        cs = 32
        out = torch.zeros(orig_dim)

        for pkt in sorted(packets, key=lambda p: p.chunk_idx):
            i = pkt.chunk_idx
            rht = self._get_rht(pkt.chunk_size, i)
            qz_r = LloydMaxQuantizer(pkt.n_bits_r).default_fit(scale=3.0)
            qz_theta = self._get_qz(pkt.n_bits_theta)

            r_rec = qz_r.dequantize(torch.tensor([pkt.r_idx]))
            theta_rec = qz_theta.dequantize(torch.tensor(pkt.theta_indices))
            v_polar = polar_inverse(r_rec, theta_rec.unsqueeze(0)).squeeze(0)
            v_chunk = rht.inverse(v_polar.unsqueeze(0)).squeeze(0)

            # QJL correction
            qjl_signs = _unpack_bits(pkt.qjl_bits, self.qjl.proj_dim)
            correction = self.qjl.decode(qjl_signs, pkt.chunk_size)
            v_chunk = v_chunk + correction

            start = i * cs
            end = start + pkt.chunk_size
            out[start:end] = v_chunk

        return out


# -------------------------------------------------------------------------
# Bit packing utilities
# -------------------------------------------------------------------------

def _pack_bits(bool_tensor: torch.Tensor) -> bytes:
    """Pack a bool tensor into bytes."""
    flat = bool_tensor.flatten().tolist()
    n = len(flat)
    n_bytes = math.ceil(n / 8)
    ba = bytearray(n_bytes)
    for i, b in enumerate(flat):
        if b:
            ba[i // 8] |= (1 << (i % 8))
    return bytes(ba)


def _unpack_bits(data: bytes, n_bits: int) -> torch.Tensor:
    """Unpack bytes back to bool tensor of length n_bits."""
    out = []
    for i in range(n_bits):
        byte_idx = i // 8
        bit_idx = i % 8
        if byte_idx < len(data):
            out.append(bool((data[byte_idx] >> bit_idx) & 1))
        else:
            out.append(False)
    return torch.tensor(out, dtype=torch.bool)
